Initial Developer Documentation (#148)
10up/plugin-check: docs/technical-overview.md

# Technical Overview

> The Plugin Checker works by running checks against a single plugin. Each check will test for a specific issue and raise either an error or warning depending on the severity.

Reviewer: ```suggestion
The Plugin Checker works by executing checks against a single plugin. Each check will test for a specific issue and raise either an error or a warning depending on the severity.
```

> The Plugin Checker performs these checks following the process below.

Reviewer: ```suggestion
The Plugin Checker performs these checks by following the process below:
```

There is a very important grammatical difference between these two. By not adding the `by` prior to `following`, it implies that the plugin performs the checks _after_ the process below, whereas by adding the `by` it implies that the process is directly used as the approach for performing the checks.

> 1. Determine checks to run based on the request.

Reviewer: ```suggestion
1. Determine which checks to run based on the request.
```

By connecting "which" to "checks", it provides more clarity by explicitly indicating that the process needs to determine the specific checks to be run based on the request.

The first sentence is still grammatically correct & acceptable, but slightly ambiguous.

> 2. Run any check preparations required to prepare the environment.

Reviewer: ```suggestion
2. Execute necessary check preparations to set up the environment.
```

I think it's unnecessary to state something along the lines of "preparations required to prepare" as a preparation is exactly that. It's like saying "reversing backwards".

> 3. Run all checks against the plugin.

Reviewer: ```suggestion
3. Execute a comprehensive set of checks against the plugin.
```

I'm going to play devils advocate here, but based on the query (or based on future additions) we might not be running _all_ checks against a plugin, but rather a selection or set of tests.

The sentence can be simplified further if desired, if all checks do indeed run:

"Execute all checks against the plugin.", or we can just keep it as is if that is the case.

> 4. Report all errors and warnings raised by the checks run.

Reviewer: ```suggestion
4. Ensure all errors and warnings raised during the checks are reported.
```

This change provides a clearer, more active instruction. By using a word like "ensure", we emphasize the importance of reporting the errors and warnings.

## Checks

> Checks are single PHP classes that run code to test a plugin for a specific problem.

Reviewer: ```suggestion
Checks are individual PHP classes designed to test a plugin for a specific problem when executed.
```

Better flow, easier to read.

> Every check contains a `run()` method which accepts an instance of the `Check_Result` class. The `Check_Result` instance is used to add any errors or warnings raised by the check to be reported later.

Reviewer: ```suggestion
Each check contains a `run()` method that accepts an instance of the `Check_Result` class. This `Check_Result` instance is used to collect and report any errors or warnings raised by the check.
```

> Currently, static and runtime checks are the 2 types of checks that can be run by the Plugin Checker.

Reviewer: ```suggestion
The Plugin Checker currently supports two types of checks: static and runtime checks.
```

Simplification, easier to read.

### Static Checks

> Static checks are used to perform tests against the codebase without running any code. This is similar to other static analysis tools such as PHP Code Sniffer.

Reviewer: ```suggestion
Static checks are used to perform tests against the codebase without running any code, similar to other static analysis tools like PHP Code Sniffer.
```

Combining the two sentences for simplicity.

> Static checks can run existing PHPCodeSniffer sniffs, such as those in the [WordPress Coding Standards](https://github.com/WordPress/WordPress-Coding-Standards), or they can be implemented in a more manual way, e.g. searching for specific patterns across all files, similar to how the checks of the [WordPress Theme Check tool](https://github.com/WordPress/theme-check) are implemented.

Reviewer: ```suggestion
Static checks can utilize existing PHPCodeSniffer sniffs, such as those in the [WordPress Coding Standards](https://github.com/WordPress/WordPress-Coding-Standards), or they can be implemented manually by searching for specific patterns across all files, similar to the checks in the [WordPress Theme Check tool](https://github.com/WordPress/theme-check).
```

Slight simplification and better flow.

Static checks implement the `Static_Check` interface.

### Runtime Checks

> Runtime checks perform tests by executing the code within a test environment. The plugin code is required to run in order to test a specific output.

Reviewer: ```suggestion
Runtime checks perform tests by executing the plugin code within a dedicated test environment, allowing for verification of specific outputs.
```

Simplification and better flow. By adding "dedicated test environment" we clarify exactly where the plugin code is executed when performing checks.

> As runtime checks execute code against a test environment they often include preparations in order to prepare the runtime environment with required configuration ahead of check being performed.

Reviewer: ```suggestion
Runtime checks execute code within a test environment, often requiring preparations to configure the runtime environment with the necessary configuration prior to performing the checks.
```

Improved readability, clarity and flow.

Runtime checks implement the `Runtime_Check` interface.

### Check Stabilities

> Every check is assigned a stability value so the correct checks are run for each use case. There are currently 2 stability values, Stable and Experimental.

Reviewer: ```suggestion
Each check is assigned a stability value to ensure appropriate check selection based on specific use cases. Currently, there are two stability values: Stable and Experimental.
```

Enhanced clarity, more coherent.

> Checks are a assigned a stability value using either the `Stable_Check` or `Experimental_Check` trait. Checks with the `Stable_Check` trait are always run by both the CLI and WordPress admin screens. Experimental checks are only run via the CLI with the `--include-experimental` flag.

Reviewer: ```suggestion
Checks are assigned a stability value using either the `Stable_Check` or `Experimental_Check` trait. Those with the `Stable_Check` trait are always executed on both the CLI and WordPress admin screens, whereas Experimental checks are exclusively run via the CLI using the `--include-experimental` flag.
```

Better flow and clarity, easier readability.

### Check Categories

> Check categories allow checks to be run for specific use cases. All checks require a category which is set by implementing a `get_categories` method within the check class. This method returns an array containing one or more categories the check belongs to. The categories set should be made up by the constants found in the `Check_Categories` class rather than setting them as strings.

Reviewer: ```suggestion
Check categories enable the execution of checks for specific use cases. Each check requires a category, which is determined by implementing a `get_categories` method within the check class. This method returns an array containing one or more categories to which the check belongs. The categories should be specified using the constants found in the `Check_Categories` class, rather than setting them as strings.
```

More concise, clearer phrasing, better flow.

### Preparations

> Preparations are used to prepare the test environment ahead of running a runtime check to ensure that they run correctly.

Reviewer: ```suggestion
Preparations are utilized to set up the test environment before running a runtime check, ensuring their successful execution.
```

Better flow.

> Preparations can include any logic from adding filters, or creating test content to perform checks against.

Reviewer: ```suggestion
Preparations can encompass various tasks, such as adding filters or creating test content, which are performed to enable comprehensive checks.
```

Indicate a broader range of tasks that could be part of the preparation. Further clarifies the purpose of the preparations.

## The Test Environment

> In the context of runtime checks, the Plugin Checker ensures that the checks are performed in a controlled environment separate from the production site. This approach prevents any unintended changes to the actual WordPress site.

Reviewer: ```suggestion
In the context of runtime checks, the Plugin Checker ensures that the checks are executed within a controlled environment separate from the production site, thus preventing any unintended changes to the actual WordPress site.
```

Combining the two sentences for better/easier flow. Using "executed" as opposed to "performed" given that the tests are methods that are ultimately executed.

> To achieve this, the Plugin Checker employs the following practices:

Reviewer: ```suggestion
To achieve this, the Plugin Checker implements the following practices:
```

By using "implements" instead of "employs", we make it very clear that the Plugin Checker actively applies the practices that follow.

> * Separate Database Tables: During runtime checks, a distinct set of database tables is utilized. This isolation ensures that the checks do not interfere with the data of the production site.

Reviewer: ```suggestion
* Separate Database Tables: During runtime checks, the Plugin Checker utilizes a distinct set of database tables. This isolation ensures that the checks do not interfere with the data of the production site.
```

By clarifying that it is the _Plugin Checker_ that utilizes the distinct set of db tables, we remove any potential ambiguity.

> * Restricted Plugin Activation: Only the plugin being tested is activated in the runtime environment. By keeping other plugins deactivated, the checks focus solely on the plugin under examination.

Reviewer: ```suggestion
* Restricted Plugin Activation: In the runtime environment, only the plugin being tested is activated. By deactivating other plugins, the checks focus solely on the plugin under examination.
```

> It's important to note that while these measures aim to minimize the impact on the WordPress site, it is strongly advised not to perform runtime checks using the Plugin Checker on a production site. Despite the precautions taken, there is still a possibility of unintended consequences or conflicts.

Reviewer: ```suggestion
It is important to note that while these measures aim to minimize the impact on the WordPress site, it is strongly advised not to perform runtime checks using the Plugin Checker on a production site. Despite the precautions taken, there remains a possibility of unintended consequences or conflicts.
```

Slightly better flow.

[Back to overview](./README.md)

# Technical Overview

The Plugin Checker works by executing checks against a single plugin. Each check will test for a specific issue and raise either an error or a warning depending on the severity.

The Plugin Checker performs these checks by following the process below:

1. Determine which checks to run based on the request.
2. Execute necessary check preparations to set up the environment.
3. Execute a comprehensive set of checks against the plugin.
4. Ensure all errors and warnings raised during the checks are reported.

## Checks
Checks are individual PHP classes designed to test a plugin for a specific problem when executed.

Each check contains a `run()` method that accepts an instance of the `Check_Result` class. This `Check_Result` instance is used to collect and report any errors or warnings raised by the check.

The Plugin Checker currently supports two types of checks: static and runtime checks.

### Static Checks

> Static checks are used to perform tests against the codebase without running any code, similar to other static analysis tools like PHP Code Sniffer.

Reviewer: ```suggestion
Static checks are used to perform tests against the codebase without running any code within the plugin being checked, similar to other static analysis tools like PHP Code Sniffer.
```

I know this may seem like a nit pick suggestion, but reading it prior when I read "perform tests against the codebase without running any code" it sounded like the plugin checker itself did not run any code, but of course it does :) I just added this to clarify that the plugin being checked has no code executed.

Static checks implement the `Static_Check` interface.

### Runtime Checks

Runtime checks perform tests by executing the plugin code within a dedicated test environment, allowing for verification of specific outputs.

Runtime checks execute code within a test environment, often requiring preparations to configure the runtime environment with the necessary configuration prior to performing the checks.

Runtime checks implement the `Runtime_Check` interface.

### Check Stabilities

Each check is assigned a stability value to ensure appropriate check selection based on specific use cases. Currently, there are two stability values: Stable and Experimental.

Checks are assigned a stability value using either the `Stable_Check` or `Experimental_Check` trait. Those with the `Stable_Check` trait are always executed on both the CLI and WordPress admin screens, whereas Experimental checks are exclusively run via the CLI using the `--include-experimental` flag.

### Check Categories

> Check categories enable the execution of checks for specific use cases. Each check requires a category, which is determined by implementing a `get_categories` method within the check class. This method returns an array containing one or more categories to which the check belongs. The categories should be specified using the constants found in the `Check_Categories` class, rather than setting them as strings.

Reviewer: This is more of a comment/observation than a change request: It is stated that one should only use the constants found in the `Check_Categories` class. Is there (or should there be) a means of extending the categories? If so, we should reference the filters. If not, it MIGHT be worth stating that custom categories are not (yet??) supported and adding a category that is not one of these constants, will have no effect and the check won't appear?

Reviewer 2: Thank you, @10upsimon, for your question. Currently, our plugin supports only the categories that are listed here: [Check_Categories.php](https://github.com/10up/plugin-check/blob/trunk/includes/Checker/Check_Categories.php#L17-L22). However, users can still define their own categories when registering new checks. It is important to note that, as of now, these custom categories may not be reflected in the WP admin screen. Nevertheless, users can make use of these custom categories when employing the CLI command to filter results by category if desired. We do plan to implement a filter for extending categories in the future, as and when it becomes necessary.

Reviewer: @mukeshpanchal27 I think it's important to note these important points in the docs.

Reviewer: ```suggestion
Check categories enable the execution of checks for specific use cases. Each check requires a category, which is determined by implementing a `get_categories` method within the check class. This method returns an array containing one or more categories to which the check belongs. The categories should be specified using the constants found in the `Check_Categories` class, rather than setting them as strings.

It should be noted that if categories are assigned that are _not_ part of the constants found in the `Check_Categories` class, said checks may not appear in the dashboard (unless they also contain other categories that _are_ defined in the `Check_Categories` class as constants). They will, however, be able to be surfaced via the wp-cli method of running checks, where checks assigned to any category (existing or custom) will be able to be surfaced and filtered via applicable cli arguments. 
```

> Preparations are used to prepare the test environment ahead of running a runtime check to ensure that they run correctly.
> 
> Preparations can include any logic from adding filters, or creating test content to perform checks against.
> 
> Every preparation contains a `prepare()` method, which returns a cleanup function that is called to revert the changes made by that method. Preparations can be implemented as their individual classes or alternatively as part of a specific check.

Reviewer 3: There is still quite some overlap here with the explanation in the "Creating a runtime check" article. I think especially the last sentence here goes too much into detail, so I suggest we remove it from here, as it's already explained in the other article.
```suggestion
Preparations are used to prepare the test environment ahead of running a runtime check to ensure that they run correctly.

Preparations can include any logic from adding filters, or creating test content to perform checks against.
```

Author: Updated

## The Test Environment

In the context of runtime checks, the Plugin Checker ensures that the checks are performed in a controlled environment separate from the production site. This approach prevents any unintended changes to the actual WordPress site.

To achieve this, the Plugin Checker employs the following practices:

Separate Database Tables: During runtime checks, a distinct set of database tables is utilized. This isolation ensures that the checks do not interfere with the data of the production site.

> Restricted Plugin Activation: Only the plugin being tested is activated in the runtime environment. By keeping other plugins deactivated, the checks focus solely on the plugin under examination.

Reviewer 3: For better formatting:
```suggestion
To achieve this, the Plugin Checker employs the following practices:

* Separate Database Tables: During runtime checks, a distinct set of database tables is utilized. This isolation ensures that the checks do not interfere with the data of the production site.
* Restricted Plugin Activation: Only the plugin being tested is activated in the runtime environment. By keeping other plugins deactivated, the checks focus solely on the plugin under examination.
```

Author: Thanks @felixarntz this has been updated.
